### DISCUSSION TITLE
How to write unit test for Javascript code in pb_hooks?

### DISCUSSION THREAD

**Discussion by mdnahas** [pocketbase/pocketbase#4083] [Q&A] [answered]
I'm writing complex server-side code in Typescript.  (It compiles to Javascript, which resides in the pb_hooks directory.)  I need to test the code.  I'd love to test all the code, but I don't think that's possible without compiling my own copy of PocketBase.   

Instead, I've separated the code into two parts: one that reads/writes the database and the other that computes functions on that data.  I'm trying to write unit tests on the non-database functions.  I installed Vitest, which is a Typescript/JavaScript testing framework that I'm also using on the frontend.  The problem I'm running into is that the code is not completely separated from the PocketBase DB code.  It has types like DateTime, inside records, which Vitest cannot find a definition for.

Is there a recommended way to write unit tests for Javascript server-side code for PocketBase?
Is there a way to get my database-free code to run with types like DateTime, etc.?
Can anyone point me to some examples (e.g., Github repos) using the recommended approach?

**Comment by ganigeorgiev** [comment, marked as answer]
Unit testing with the JS `pb_hooks` is not possible, unless you decide to mock every translated Go API.

External integration testing is doable by sending real requests to test PocketBase instances that operates with a test `pb_data` (_each test could start their own PocketBase process with a copy of the test `pb_data` to avoid mutations_).

But in general, if you need tests my recommendation is to use Go. We also have some test helpers for this case (see [Testing](https://pocketbase.io/docs/go-testing/)).

**Reply by mdnahas** [reply]
Ok.  Thanks for the quick reply, @ganigeorgiev !
